### ISSUE TITLE
cryptsetup/crypttab BitLocker entries are not mounted

### ISSUE THREAD

**Issue by luzat** [systemd/systemd#15057]
Starting with cryptsetup 2.3.0, cryptsetup supports BitLocker. On Debian with systemd 244 I am now using crypttab entries such as:

```
data PARTUUID=123-456-789 /etc/cryptsetup/data.key discard
```

This works fine with Debian's (and possibly upstream's) `cryptdisks_start`. Unfortunately, systemd 244 does not seem to support this and I did not see a relevant change skimming through the logs for 245. With sytemd I get:

```
systemd-cryptsetup[1136]: Failed to load LUKS superblock on device /dev/disk/by-partuuid/123-456-789: Invalid argument
```

Now, other types support specifying an explicit type (luks, plain, tcrypt, ...), but it seems that this has not been introduced for BitLocker. The format should probably be deduced automatically. #442 might be relevant there. For reference, this is what the `changelog.Debian.gz` for 2:2.3.0-1 says:

>   * New upstream release, introducing support for BitLocker-compatible
    devices (BITLK format) used in Windows systems.
    WARNING: crypttab(5) support for these devices is currently *experimental*
    and requires blkid from util-linux >=2.33 (i.e., Buster or later).  These
    devices currently have no keyword to use in the 4th field (unlike 'luks'
    or 'plain'), the device type is inferred from the signature instead.
>  * crypttab(5): Make the 4th field (options) optional so we don't have to
    introduce a new keyword for each new device type.  (That field is also
    optional in the systemd implementation.)  Other fields (dm target name,
    source device, and key file) remain required.

I think systemd should handle no type or an empty 4th field fine in this case, too. Contrary to `plain` devices I can't find a workaround in crypttab here either currently.

**Comment by mxfm** [comment]
Ping. What is status of this?

**Comment by poettering** [comment]
@mxfm To my knowledge noone is working on this. If this feature is important to you, prep a patch or find somebody to do I for you, or wait until one day someone comes along to implement this for you. It's probably not difficult, after all libcryptsetup already does all the hard work. It just needs some minor glue in systemd-cryptsetup to make available.

**Comment by mxfm** [comment]
https://github.com/systemd/systemd/pull/15979

**Comment by luzat** [comment]
@mxfm Thank you, this seems to work.

I didn't follow through completely, but the follow-up by @poettering in #16111 seems to make adding `bitlk` to `crypttab` a requirement? This is a bit unfortunate, because dm-crypt/cryptsetup do not seem to support that within crypttab:

    # cryptdisks_start foo
    Starting crypto disk...cryptsetup: WARNING: foo: ignoring unknown option 'bitlk'

So, `systemd-cryptsetup` works now, but other tools give warnings. No way around that? Not sure if this should be a bug with `systemd` or `cryptsetup`.

**Comment by mxfm** [comment]
@luzat, I don't quite understand. The follow-up made absence of any parameter (luks, plain, bitlk) equivalent to assuming luks. It did not make 'bitlk' option a requirement - this would be illogical to make windows stuff parameter to be requirement.  

Regarding the error message - from which program does it come? Is it cryptsetup? It seems cryptsetup does not support parsing crypttab by itself, so I do not know the source of the message.

**Comment by luzat** [comment]
@mxfm Sorry, I was thinking that `crypttab` support was included in `cryptsetup`, but in this case it's provided by Debian scripts. I have submitted a [bug report](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=967853) to Debian's bug tracker.

**Comment by luzat** [comment]
For reference: The latter issue has now been fixed in Debian's cryptsetup in 2:2.3.3-2.
